### ISSUE TITLE
br close persists JSONL but not SQLite — dual-store divergence on close path

### ISSUE THREAD

**Issue by JYeswak** [Dicklesworthstone/beads_rust#285]
## What happened

Running `br 0.2.5`. Calling `br close <id>` on an open issue exits 0 and writes the closed status to `.beads/issues.jsonl`, but **does not update `.beads/beads.db`**. SQLite row remains `status='open'` while JSONL row is `status='closed'`. `br show` reads JSONL and reports closed; raw `sqlite3 ... SELECT status` returns open.

The dirty tracker also doesn't notice — `br sync --flush-only` immediately after the close reports `Nothing to export (no dirty issues)`. So the JSONL→DB→JSONL reconciliation that would normally heal divergence never fires for this row, because nothing ever flagged the row as needing reconciliation.

The drift accumulates silently. On a tracker with ~1k closed events I see ~24 rows where last JSONL event is `closed` but SQLite still reports `open`.

## Repro

```bash
# Pre-flight — confirm baseline
br --version  # br 0.2.5
TARGET=<bead-id>
BACKUP=/tmp/issues-before-repro.jsonl
cp .beads/issues.jsonl "$BACKUP"
br sync --rebuild --json

# Confirm pre-state — open in both stores
br show "$TARGET"                                          # status: open
sqlite3 .beads/beads.db "SELECT status FROM issues WHERE id='$TARGET'"  # open

# The close
br --lock-timeout 10000 close "$TARGET" --reason "diagnostic repro"
# exit 0
# stdout reports "Closed <id>: <title>"

# Post-state — divergence
br show "$TARGET"                                          # status: closed (reads JSONL)
sqlite3 .beads/beads.db "SELECT status FROM issues WHERE id='$TARGET'"  # status: open (BUG)

# The dirty tracker doesn't notice
br sync --flush-only
# Nothing to export (no dirty issues)

# Restore
cp "$BACKUP" .beads/issues.jsonl
br sync --rebuild --json
```

## Expected vs observed

| | Expected | Observed |
|---|---|---|
| `br close` exit code | 0 | 0 ✓ |
| JSONL row status | closed | closed ✓ |
| SQLite row status | closed | **open** ✗ |
| Dirty tracker post-close | row marked dirty (so flush reconciles) | row NOT marked dirty |
| `br sync --flush-only` post-close | exports the dirty row | reports "Nothing to export" |

`br show` and the SQLite UPDATE-issues path disagree by store.

## File:line citations

- `src/cli/commands/close.rs:249` — close path sets `status: Some(Status::Closed)` on the in-memory update
- `src/cli/commands/close.rs:353` — call to `storage_ctx.flush_no_db_if_dirty()`
- `src/config/mod.rs:357-362` — `flush_no_db_if_dirty` returns early `if !self.no_db` (so in db-backed mode this is a no-op) AND only fires if `get_dirty_issue_count() != 0`
- `src/storage/sqlite.rs:779` — `UPDATE issues SET ... WHERE id = ?` — this is where the SQLite write would land if the close path reached it

The shape suggests the close path serializes the closed state to JSONL on its way out but the SQLite UPDATE either never fires or fires inside a transaction that doesn't commit when the JSONL export is the destination. Dirty tracking and JSONL persistence end up on different sides of the boundary.

## Why this matters

`br close` is the canonical lifecycle terminator. When it silently desyncs the two stores:

- `br ls` and `br show` disagree (one reads SQLite directly, the other reads JSONL via cache)
- Worker agents that close beads via `br close` and re-query via `br show` see "closed" — they trust it. Downstream readers querying SQLite see "open" and re-route the bead as still-actionable. Same bead gets re-dispatched, re-worked, re-closed.
- The dirty tracker no-op means `br sync --flush-only` won't heal it. The only path back to consistency is `br sync --rebuild` from JSONL — destructive on transient SQLite state — which most operators won't run unless explicitly suspected.
- The drift compounds: the more close events, the larger the gap between stores, until a rebuild is forced.

## Out of scope

- Not asking for a feature; asking for the existing close-path contract (close updates the canonical store) to hold.
- Not submitting a PR — local workaround is direct SQL `UPDATE issues SET status='closed', closed_at=datetime('now'), updated_at=datetime('now'), close_reason=?` (must satisfy the `closed_at IS NOT NULL` CHECK constraint), but this doesn't dirty-mark the row so JSONL still drifts. Canonical recovery in our hands stays `br sync --rebuild`.
- Pre-screened against #274 (Closure-time policy gates: required fields, references, closeouts, signatures, attribution) — that issue is about validation discipline at close-time, this is about persistence consistency. Different concern, non-duplicate.

**Comment by Dicklesworthstone** [comment]
Thanks for the precise repro and the divergence count (~24 stale rows / ~1k closes is exactly the kind of silent-corruption case we want to catch).

Initial triage notes from a quick walk through the close path so you have a head start:

- `br close` ultimately hits `update_issue_with_recovery` (`src/cli/commands/mod.rs:276`) which calls `retry_mutation_with_jsonl_recovery` → `storage.update_issue` (`src/storage/sqlite.rs:2217`). If `mutate()` returns `Ok`, the SQLite row should be updated under a CONCURRENT transaction.
- After the mutation, close.rs runs `flush_no_db_if_dirty()` then `auto_flush_if_enabled()` (`src/cli/commands/close.rs:1025-1033`). The latter exports SQLite → JSONL.
- The fact that the **dirty tracker** sees zero dirty rows post-close (`get_dirty_issue_count == 0`) is the smoking gun — that tracker is updated alongside successful SQLite mutations, so its silence implies the SQLite UPDATE either (a) didn't fire, (b) committed against a different row, or (c) silently rolled back. The "JSONL is closed but SQLite isn't" symptom is consistent with `retry_mutation_with_jsonl_recovery`'s fallback path writing JSONL after a swallowed mutation error.
- `Issue #245` already documented one tracker-related corruption case where `get_issue_from_conn` inside a write transaction returns zero rows even though the read path finds it; the comment at sqlite.rs:2226 references it. Worth checking whether the 24 drifted rows correlate with that path (look for warnings tagged `update_issue: row not found inside write transaction`).

Two things that would massively accelerate a fix:

1. `RUST_LOG=br::storage::sqlite=trace,br::cli::commands::close=trace br --lock-timeout 10000 close <id>` against the precise repro — the trace events from `mutate` and `update_issue_with_recovery` should show whether the SQLite write was attempted, succeeded, or got rerouted through JSONL recovery.
2. `br doctor --json` immediately after a divergence is observed — its rebuild diff would show exactly which rows the dirty tracker is missing.

Leaving open as a tracking issue while we dig in.

**Comment by JYeswak** [comment]
The `update_issue_with_recovery` → `retry_mutation_with_jsonl_recovery` → `flush_no_db_if_dirty` walkthrough is exactly what was needed for useful repro work on this side.

Both signals you asked for are tractable:

1. `RUST_LOG=br::storage::sqlite=trace,br::cli::commands::close=trace br --lock-timeout 10000 close <id>` against a controlled repro — the divergence sample on this side is roughly 24 stale rows out of ~1k closes, so a small batch of those bead IDs replayed against a snapshotted DB+JSONL pair should surface whether the JSONL-recovery fallback is firing. Will share trace excerpts once one shows the fallback path active.

2. `br doctor --json` immediately post-divergence — wireable into the close path so any close that completes without `dirty_count` decrementing triggers `br doctor --json` and dumps the rebuild diff to a per-task receipt. That gives a stable signal stream rather than one-off manual repros.

Will check the warnings tag you flagged (`update_issue: row not found inside write transaction`) against the 24-row sample; if those correlate with the `Issue #245` tracker case, we'll cite both in the trace dump.

Happy to test a fix branch once you have one — just name the branch when ready and we'll run the precise repro plus our close-path tests against it.

**Comment by Dicklesworthstone** [comment]
Reproduced the call graph: `br close` routes through `update_issue_with_recovery` (src/cli/commands/mod.rs:279), which is shared with `defer` and `reopen`. The dirty-tracker behavior you observed — `br sync --flush-only` reports `Nothing to export` immediately after a close — points at the close path skipping the `mark_dirty(issue_id)` call that the other mutation paths take (`SqliteStorage::mark_dirty` callsites at src/storage/sqlite.rs:561 etc.). Without `dirty_issues` being flagged, the JSONL→DB reconciliation never gets the signal to bring the SQLite row forward, and the divergence accumulates exactly as you describe.

Real bug. The fix shape:
1. `update_issue_with_recovery` (or `record_close_metadata` immediately after it) explicitly calls `storage.mark_dirty(id)` on the close path so the row is queued for the next flush.
2. A regression test that round-trips `br close` → `br sync --flush-only` and asserts SQLite reports `status='closed'` after the flush (the existing close-path unit tests assert JSONL but not SQLite).
3. Investigate whether `update_issue_with_recovery` is silently swallowing an error from the SQLite update — the JSONL-but-not-DB shape suggests JSONL is being written from an in-memory intent before the SQLite write actually lands.

Leaving open; will reference the fix commit when it lands. The 24-row divergence on your tracker is recoverable via `br sync --rebuild`, but that doesn't address the silent close path.

**Comment by Dicklesworthstone** [comment]
Correction to my earlier triage: re-audited the close path on current main against your symptom shape and found that `mark_dirty` **is** being called.

Call graph:
- `br close` → `update_issue_with_recovery` (cli/commands/mod.rs:279) → `storage.update_issue` (sqlite.rs:2223) → `ctx.mark_dirty(id)` (sqlite.rs:2634).

Pinned the contract that's currently in place with regression test [59bd2fc4](https://github.com/Dicklesworthstone/beads_rust/commit/59bd2fc4) (`test_close_path_marks_dirty_and_persists_to_db`): close-as-update writes `status='closed'` to the SQLite store AND enqueues the row in `dirty_issues`. If anyone regresses either half, the test fails fast.

So the 2.4% drift you observed isn't explained by my original hypothesis. What I still need to chase:

1. Does `br 0.2.5` specifically (the binary you ran) have a different code path than current main? Worth checking `git log v0.2.4..v0.2.5 -- src/storage/sqlite.rs src/cli/commands/close.rs` to see if any close-path commits landed between releases.
2. Is the divergence entering through a non-close write path? `update_status`, `record_close_metadata`, or auto-flush after a partial transaction commit. `record_close_metadata` is interesting: it runs *after* the close commits and any error is logged-and-swallowed — but that path writes its own row, not the issues row, so it shouldn't cause status drift on its own.
3. Is the drift specifically on the close path, or also on other status transitions? Your 24/1000 rate is consistent with a concurrency edge case — two writers racing the same row, one of which hits a retry path that swallows an error.

Could you share: (a) the `source_repo` in one of the 24 divergent rows (we want to know if they cluster by project), (b) whether `br doctor` on the affected DB reports any `integrity_check` findings or stale partial indexes (related to #288), (c) a small slice of the activity log around one of the divergent closes if `.beads/.br_history/` is enabled?

Leaving open — the diagnostic test is in place, but the actual root cause needs the additional context.

**Comment by Dicklesworthstone** [comment]
Status update: investigation continues. The regression test pinned the contract that IS in place ([59bd2fc4](https://github.com/Dicklesworthstone/beads_rust/commit/59bd2fc4)) — close path marks dirty on current main. The 2.4% drift you observed in 0.2.5 either (a) was a 0.2.5-specific bug fixed between versions, OR (b) enters through a different mechanism we haven't reproduced yet.

To narrow it down without your DB on hand, I would need:
1. **The exact `br --version` output** of the binary you reproduced this on. If 0.2.6+, the bug is still live and we need to keep digging.
2. **The 24-row sample**: `source_repo`, IDs, and a `br doctor --json` output against the affected DB. The id-set divergence surfacer ([258328ae](https://github.com/Dicklesworthstone/beads_rust/commit/258328ae)) for #286 lands in the next release; if you can rerun `br doctor` once that ships, the `db_jsonl_id_set_mismatch` anomaly will tell us if the same rows show only_db or only_jsonl.
3. **One concrete close that produced drift**: the actor field on the close, what other br commands ran in the same session, and whether `.beads/.br_history/` is enabled. Concurrency edge cases are the most plausible explanation given the 2.4% rate.

Leaving open. Will move to closed with reference-to-fix-commit once we have a reproducer.

**Comment by Dicklesworthstone** [comment]
Status update: v0.2.7 has shipped, and I want to ask whether the drift is still observable on a fresh binary.

The audit at 59bd2fc4 confirmed the contract — `br close` → `update_issue_with_recovery` → `storage.update_issue` → `ctx.mark_dirty(id)` — is fully wired on current main, and a regression test now pins that path so a future refactor that drops the `mark_dirty` call (the hypothesized root cause) fails fast against the exact shape this issue reports. So the missing-mark-dirty story doesn't match the code that v0.2.7 actually ships, and I haven't been able to reproduce the divergence symptom against it.

But the 2.4% drift you observed on a real tracker is real signal — the file:line citation pointing at `flush_no_db_if_dirty` returning early in `!self.no_db` mode is also accurate today, so there may be a different code path that still skips the SQLite write in some race I haven't caught. If you can:

1. Upgrade to v0.2.7
2. Re-run the close path on a row where you'd previously observed drift
3. Confirm whether the SQLite/JSONL stores diverge

…then either we'll have a clean confirmation that the existing path is correct on a fresh binary (and the drift on your tracker came from a pre-v0.2.7 close ledger), or we'll have a fresh failure to bisect against the audit.

If you want I can also add a startup self-check that walks the closed-event tail and reconciles store-to-store divergence so the existing 24 drifted rows on your tracker get healed without needing a manual `br sync --rebuild`. Let me know if that's useful and I'll prioritize it.

**Comment by Dicklesworthstone** [comment]
Reproduced and verified fixed in br 0.2.6.

Test (against br 0.2.6 binary on a fresh init):

\`\`\`
$ br create --title \"test issue for #285\" --type bug --priority 2
br-test-...-8p4

$ sqlite3 .beads/beads.db \"SELECT status FROM issues WHERE id='br-test-...-8p4'\"
open

$ br close br-test-...-8p4 --reason \"diagnostic for #285\"
✓ Closed br-test-...-8p4

$ sqlite3 .beads/beads.db \"SELECT status FROM issues WHERE id='br-test-...-8p4'\"
closed  ✓
\`\`\`

The repro you posted in 0.2.5 now exits with both stores in agreement after \`br close\`. SQLite shows \`closed\` immediately.

What's covered now:
- 59bd2fc4 — pins the dirty-tracker contract in a regression test so a regression of \"close writes JSONL but not SQLite\" or \"close doesn't mark the row dirty\" surfaces in CI rather than as silent drift.
- The current close path goes \`br close → update_issue_with_recovery → storage.update_issue → ctx.mark_dirty(id)\`, all on the same transactional boundary — the missing-mark-dirty hypothesis from your 0.2.5 trace can't fire in this shape.

If you still see divergence on 0.2.6 or later, please reopen with the binary version and the post-close \`sqlite3 .beads/beads.db \"SELECT status FROM issues WHERE id=?\"\` output — the diagnostic shell snippet from your repro is exactly the right thing to attach, so don't change a thing on that front.

Thanks for the file:line citations + paired before/after table; both made this fast to verify and pin.
